src/app/mesh/qgsmaptooleditmeshframe.h

```diff
@@ -0,0 +1,180 @@
+    void setZValue( double z );
+
+    /**
+     *  Sets the current value of the widget and set it as the default one ,
+     *  that is the value that is retrieve if the z value spin box is cleared
+     */
+    void setDefaultValue( double z );
+
+    /**
+     *  Installs an event filter (see QObject::eventFilter()) on the Z value spin box.
+     *  This \a filter can be used to control keyboard entry when the focus is on the Z value spin box.
+     */
+    void setEventFilterOnValueSpinbox( QObject *filter );
+
+  private:
+    QgsDoubleSpinBox *mZValueSpinBox = nullptr;
+    double mValue;
+};
+
+class APP_EXPORT QgsMapToolEditMeshFrame : public QgsMapToolAdvancedDigitizing
+{
+    Q_OBJECT
+  public:
+    //! Constructor
+    QgsMapToolEditMeshFrame( QgsMapCanvas *canvas );
+    ~QgsMapToolEditMeshFrame();
+
+    void deactivate() override;
+    void activate() override;
+    bool populateContextMenuWithEvent( QMenu *menu, QgsMapMouseEvent *event ) override;
+    Flags flags() const override;
+
+  protected:
+    bool eventFilter( QObject *obj, QEvent *ev ) override;
+    void cadCanvasPressEvent( QgsMapMouseEvent *e ) override;
+    void cadCanvasReleaseEvent( QgsMapMouseEvent *e ) override;
+    void cadCanvasMoveEvent( QgsMapMouseEvent *e ) override;
+    void keyPressEvent( QKeyEvent *e ) override;
+    void keyReleaseEvent( QKeyEvent *e ) override;
+
+    //! Start addition of a new vertex on double-click
+    void canvasDoubleClickEvent( QgsMapMouseEvent *e ) override;
+
+  private slots:
+    void setCurrentLayer( QgsMapLayer *layer );
+    void onEdit();
+    void onEdidingStarted();
+    void onEditingStopped();
+
+  private:
+    // methods
+    const QgsMeshVertex mapVertex( int index ) const;
+    const QgsMeshFace nativeFace( int index ) const;
+
+    void highlightCurrentHoveredFace( const QgsPointXY &mapPoint );
+    void highlightCloseVertex( const QgsPointXY &mapPoint );
+
+    void createZValueWidget();
+    void deleteZvalueWidget();
+    void clear();
+    void addVertex( const QgsPointXY &mapPoint, const QgsPointLocator::Match &mapPointMatch );
+    void updateFreeVertices();
+
+    //! Checks if we are closed to a vertex, if yes return the index of the vertex;
+    int closeVertex( const QgsPointXY &point ) const;
+
+    QgsPointSequence nativeFaceGeometry( int faceIndex ) const;
+
+    QgsPointXY newFaceMarkerPosition( int vertexIndex );
+
+    void addVertexToFaceCanditate( int vertexIndex );
+    bool testNewVertexInFaceCanditate( int vertexIndex );
+
+    // selection private method
+    void setSelectedVertex( const QList<int> newSelectedVertex, bool ctrl );
+    void clearSelectedvertex();
+    void removeSelectedVerticesFromMesh();
+    void removeSelectedVerticesFromMesh( bool fillHole );
+    void selectVerticesInGeometry( const QgsGeometry &geometry, bool ctrl );
+
+    void applyZValueOnSelectedVertices();
+
+    // members
+    enum State
+    {
+      Default,
```

Comment:
not sure if Default is good name, do you mean it is before editing? or read-mode?